http4k-format/core/src/main/kotlin/org/http4k/format/AutoContentNegotiatior.kt

```diff
@@ -0,0 +1,41 @@
+package org.http4k.format
+
+import org.http4k.core.Accept
+import org.http4k.core.HttpMessage
+import org.http4k.core.Request
+import org.http4k.lens.BiDiBodyLens
+import org.http4k.lens.BodyLens
+import org.http4k.lens.ContentNegotiation
+import org.http4k.lens.Header
+import org.http4k.lens.LensExtractor
+
+class AutoContentNegotiator<T>(
+    private val defaultLens: BiDiBodyLens<T>,
+    private val alternateLenses: List<BiDiBodyLens<T>>
+): Iterable<BiDiBodyLens<T>>, LensExtractor<HttpMessage, T> {
+
+    override fun iterator() = listOf(defaultLens).plus(alternateLenses).iterator()
+
+    operator fun invoke(accept: Accept?): BiDiBodyLens<T>
+        = find { accept == null || accept.accepts(it.contentType) } ?: defaultLens
+
+    override fun invoke(target: HttpMessage): T = Header
+        .CONTENT_TYPE(target)
+        ?.let { contentType -> Accept(listOf(contentType), emptyList()) }
+        .let { accept -> invoke(accept) }
+        .invoke(target)
+
+    fun outbound(request: Request): BiDiBodyLens<T> = invoke(Header.ACCEPT(request))
+
+    fun toBodyLens() = BodyLens(
+        metas = defaultLens.metas + alternateLenses.flatMap { it.metas },
+        contentType = defaultLens.contentType,
+        getLens = this::invoke
```

Comment:
The `getLens` override will allow the preflight check to use the appropriate lens.  I'm not sure if appending all of the metas is the right thing to do here.

Reply:
Metas are used for a couple of things - 

1. to generate error messages when there are things missing or invalid from a request
2. to generate the OpenApi schema for form bodies.

So I think you're ok for this TBH.